### DISCUSSION TITLE
Static Site Generation

### DISCUSSION THREAD

**Discussion by Akryum** [nuxt/nuxt#16153] [General]
Nuxt 3 currently doesn't support SSG with `nuxt generate` (command not implemented yet).

**Comment by pi0** [comment]
Hi @Akryum. We support Nuxt 2  style (full)static generation up to bridge also with nitro.

For Nuxt 3, while it is not implemented yet, it will be finally an extension of `nuxt build` to specific caching strategies for routes (RFC: https://github.com/nuxt/framework/discussions/560). This way we can use single command for hybrid static websites so there would be no separate `nuxt generate` command but via configuration to specify some or all routes being static.

**Reply by shainegordon** [reply]
great, as going to post a similar question

I assume that when this is implemented, we will be able to just say "make all my routes static", so that for example we can use nuxt 3 (not bridge) in new projects, and deploy to S3 for example

**Reply by SherinBloemendaal** [reply]
Is there any known release date for this?

**Reply by eliawk** [reply]
any news about the released day?

**Comment by n4an** [comment]
next decide what to do (SSG/SSR/ISR page) without tell to it.

may be use of different hooks for SSG/SSG/ISR

**Reply by n4an** [reply]
also will be great to generate site and when start it to return not only html pages but also api calls result, to be able to update page on client side.
that will be hybrid mode, not full generate

**Comment by Flawe** [comment]
I would also like to chime in and ask where static generation is on the roadmap and when we might expect to be able to build on it?

**Reply by GreyXor** [reply]
try preset: "browser"

**Reply by ilyary** [reply]
I ran "NITRO_PRESET=browser npx nuxt build" on "Nuxt 3 Getting started" app
And got ✔ Server built                                                                                                     12:19:42
  └─ .output/public/_server/index.mjs (299 kB) (104 kB gzip)
So, the index.html and *.mjs files were generated but it fails to render a page on static hosting. The browser shows only "Loading..."

**Reply by Miguelklappes** [reply]
It seems like yet another black box where they are not being transparent, again.

**Reply by morozig** [reply]
> I ran "NITRO_PRESET=browser npx nuxt build" on "Nuxt 3 Getting started" app And got ✔ Server built 12:19:42 └─ .output/public/_server/index.mjs (299 kB) (104 kB gzip) So, the index.html and *.mjs files were generated but it fails to render a page on static hosting. The browser shows only "Loading..."

It seems like currently preset: "browser" produces static files with service worker and .mjs modules. So your static server needs to set [correct](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#aside_%E2%80%94_.mjs_versus_.js) MIME type for these files to work. It might be the reason for "Loading..." in your browser, check the developer console. Note however that even if service worker solution does work for you it is not SEO friendly, so most users probably will wait for a proper release.

Here is [my example](https://github.com/morozig/octagan) of Nuxt 3 statically built with service worker and hosted via Github Pages.

**Reply by ilyary** [reply]
@morozig, there is no build output in you repo, so nothing to compare

all seems to be fine
/sw.js and /_server/index.mjs are properly loaded with
Content-Type: application/javascript; charset=UTF-8

/sw.js runs its code but then 'event.target.state' is 'redundant' on mjs registration:
```
registration.active.addEventListener('statechange', (event) => {
      if (event.target.state === 'activated') {
        window.location.reload()
      }
    })
```

full index page is like this
```
<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <link rel="prefetch" href="/sw.js">
  <link rel="prefetch" href="/_server/index.mjs">
  <script>
  async function register () {
    const registration = await navigator.serviceWorker.register('/sw.js')
    await navigator.serviceWorker.ready
    registration.active.addEventListener('statechange', (event) => {
      if (event.target.state === 'activated') {
        window.location.reload()
      }
    })
  }
  if (location.hostname !== 'localhost' && location.protocol === 'http:') {
    location.replace(location.href.replace('http://', 'https://'))
  } else {
    register()
  }
  </script>
</head>

<body>
  Loading...
</body>

</html>
```

**Reply by morozig** [reply]
Yeah, now I see what you mean, but unfortunately I don't know how to fix this, maybe someone else will run into the same issue. Also [mine file](https://github.com/morozig/octagan/blob/gh-pages/index.html) looks exactly the same. But can you try to test it again from a clean starter app? Maybe it's an environment issue, because for me it just worked out of the box, I've got Windows with node v16.13.2, but the build and deployment for my repo also worked with github actions inside an ubuntu virtual machine. Here is what I did:
```
npx nuxi init nuxt3-app
cd nuxt3-app
npm i
```
Change nuxt.config.ts:
```
import { defineNuxtConfig } from 'nuxt3'

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
  nitro: {
    preset: 'browser',
  }
})
```
```
npm run build
npx http-server .output/public
```
Finally, try to open localhost:8080. If it still doesn't work then you can at least try github actions yourself.

**Reply by ilyary** [reply]
thank you for help, it really works

my problem was that I called http://localhost:8080/index.html not just http://localhost:8080/
and in this case the magic of navigator.serviceWorker.register was not working

and 'npx http-server .output/public' is a nice shortcut, I've been using bare express before


PS

Keep waiting for real SSR release))

**Comment by Chorany** [comment]
I also have a Nuxt (static site  generation) project that we are about to commence work on and would like to know what the status is. Has anyone heard anything?

**Comment by yassilah** [comment]
Hi, I'm also using Nuxt bridge to generate a full static website and though it does generate all the pages, navigation through NuxtLink is broken as payloads aren't being loaded before navigation starts. Is there some documentation on the topic?

**Reply by danielroe** [reply]
That seems like a bug. Could you raise an issue with a minimal reproduction?

**Reply by yassilah** [reply]
I'll raise an issue but reproduction is more difficult as the static site actually needs to be deployed to see the behavior i'm describing. Here's a brief summary:
- payload files aren't generated at all, i needed to add some hooks during generation process to make that happen;
- i can see that the payloads are being fetched once the website is deployed, but they are only fetched AFTER the navigation has started; I believe the expected behavior would be to start fetching, wait for the promise to resolve and then navigate when using NuxtLink;
- i just found out about the nitro preset "browser"; tried it but navigating between pages keeps trying to make API calls;

**Comment by microHoffman** [comment]
Hey, any news or info about when the SSG should become available in Nuxt 3?:) Thanks!

**Reply by reinoldus** [reply]
According to the roadmap: https://v3.nuxtjs.org/community/roadmap/ it should be released in Q1 so this month.

**Comment by Cersin** [comment]
Hello, any new estimate time for ssg? i really need it

**Reply by Miguelklappes** [reply]
Q2 2022, they always estimate it wrong.

**Reply by rhengles** [reply]
In fairness, [software estimation is hard](https://jacobian.org/2021/may/20/estimation/).

**Comment by joewagner** [comment]
Seems like this is really close to being available.  Anyone have a work around until then?  I need this by Friday 😬

**Reply by justinwride** [reply]
I too would love for static site generation to be available in Nuxt 3. Since we rely on SSG, we are continuing to use Nuxt 2.

**Comment by renatosaz** [comment]
Nothing yet?

**Reply by joewagner** [reply]
It looks like Tomorrow there is a planned release that will have this.

**Comment by timvandaatselaar** [comment]
Does anyone have this working yet? I can't get the static output to show up right.

**Reply by renatosaz** [reply]
Any news regarding this matter?   Can't we generate Static Site yet?

**Comment by joewagner** [comment]
This seems to work now. Thanks to everyone who put time into this! https://v3.nuxtjs.org/api/commands/generate

**Reply by renatosaz** [reply]
I am trying to use this feature but it seems there is a bug: 

Node version: 16.14.1

Package import specifier "#internal/nitro" is not defined in packageis not defined in package C:\projects\nuxt3\package.json imported from C:\projects\nuxt3\.nuxt\dist\server\server.mjs     
  at new NodeError (node:internal/errors:371:5)
  at throwImportNotDefined (node:internal/modules/esm/resolve:442:9)
  at packageImportsResolve (node:internal/modules/esm/resolve:819:3)
  at moduleResolve (node:internal/modules/esm/resolve:973:21)
  at defaultResolve (node:internal/modules/esm/resolve:1080:11)
  at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
  at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
  at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
  at link (node:internal/modules/esm/module_job:78:36)

**Reply by joewagner** [reply]
I got it to work with Node version 16.13.2, hopefully that's helpful

**Reply by renatosaz** [reply]
I believe it s not related. Still same error

**Reply by mimikn** [reply]
There is a related issue about that : #4555 
It may be Windows related because I'm also having the same bug and I'm on Windows 10 / Node 16.12.0.

**Reply by renatosaz** [reply]
Yes. It is a bug on windows. I was able to run on linux.  Just confirming.

**Reply by riddla** [reply]
@joewagner, the current nuxt3 RC1 version (`3.0.0-rc.1-27524767.4826918`) does not generate static HTML files for me in `.output/public`. How did you make it work?

**Reply by MarioSebastian** [reply]
in your nuxt.config, set ssr to false and target to static, then run nuxt generate. It should generate the static html files in the output folder. Here is my nuxt.config.
```
target: 'static',
ssr: false,
generate: {
    crawler: true
},
```

**Reply by joewagner** [reply]
@riddla I used the config #owl-fullDev describes, but without the `crawler` property.

**Reply by MarioSebastian** [reply]
yeah keep in mind dynamic links are still broken and do not work yet, their roadmap says sometime in Q2 2022 they will release static generation for dynamic routes. I was testing generate to see if it will crawl links and generate the dynamic pages like in nuxt 2.

**Reply by ricky11** [reply]
@owl-fullDev thanks , are dynamic links working with generate now in sept of 2022? for me, routes are generating, but none that belong to a dynamic route.  seems like most people who use nuxt3 are doing so for dynamic pages, it seems strange why generating dynamic routes have been put aside for so long.

**Reply by MarioSebastian** [reply]
@ricky11  nope, full static is still not completed, you will have to manually specify all your dynamic routes if you want to pre-render.

**Reply by ricky11** [reply]
@owl-fullDev Thanks, i dont minds to specify, but does it work for dynamic routes? How would i specific the dynamic route, would it be similar to the pages routing with the [] bracket notation?

**Comment by cetchells** [comment]
Just looking at porting a project from nuxt2. How are payloads going to be handled? So passing in a payload to a route. The payload being content from a headless CMS for example.

**Reply by cetchells** [reply]
Might be partly answering my own question, but I take it the idea is to use the hooks -> https://github.com/nuxt/framework/blob/main/packages/schema/src/types/hooks.ts#L157
So have a module handle getting the payload from 3rd party api, and it gets passed into a page on generate?

**Reply by surgiie** [reply]
Im searching for a way to do this.

In nuxt 2 you could specify the payload during a `nuxt generate`


```
routes.push({
    route: '/dynamic',
    payload: { 
        ...
    }
})
```

This doesnt appear to work anymore.

**Comment by renatosaz** [comment]
Just confirming that I was able to build the  application on linux ubuntu 20 perfectly. It generated all the routes and SSR!  On windows It only works with 'SSR: false' option .
So it is a bug on Windows.

**Reply by SebbeJohansson** [reply]
I can also confirm. Seems like this is still not working.
Running:
- Node Version:     `v17.8.0`
- Nuxt CLI v3.0.0-rc.3

**Reply by ricky11** [reply]
does it work with dynamic routes? using navigateTo?

**Comment by mirumirumi** [comment]
Can SSR and SSG be used together at the same time?
Next.js allows for customization on a page-by-page (or component-by-component) basis.

If the ISR that plan to implement can achieve something similar, I would be fine with that.

**Reply by danielroe** [reply]
Yes, they can. Even now you can use `nitro.prerender.routes` to prerender _some_ pages while SSR-ing others.

**Reply by mirumirumi** [reply]
I see!
I thought `nitro.prerender.routes` only SSG the specified routes and has no other effect. I'll give it a try :)

**Comment by ricky11** [comment]
So just to recap to enable SSG with a nuxt 3 app

quick summary of nuxt.config.js

nuxt.config.js {
nitro : {
  <anything specific to put here>
},

ssr: false,
target : 'static'

}

Now when running `npm run generate`  I can see nitro doing its magic with pre-rendering, fetching content, saving them in the paylod.js file. But certain DYNAMIC routes are not generated, this one in particular, which are all dynamic paths :

It works in dev mode, but thats because js is fetching data. 
```
  navigateTo({name:'products-category-theme-slug', params: {
    category : route.params.category,
    theme : product.attributes.theme,
    slug : product.attributes.slug
  }})
```

So finally, is the nuxt.config.js right? And what about dynamic routes?

**Reply by TheAlexLichter** [reply]
PSA: You don't need ` target: 'static'` anymore.

If you use `ssr: false`, you won't get any data in your generated site but more of a typical SPA.

**Reply by ricky11** [reply]
do you mean if we are 'npm run generate' we don't need either static nor ssr false? So setting ssr; true would make sense, or is it defaulted to true when we generate?

For the life of me, i never noticed any difference with those options when 'generate'

**Reply by TheAlexLichter** [reply]
Exactly, it is on `true` by default and you don't need to change anything when using SSG and want the content in the HTML ("typical" static site generation). `ssr: true` means the content will be rendered by the server. `nuxi generate` -> during build time ☺️

**Reply by davidhellmann** [reply]
It needs much more clearer documentation here I think.

```js
  nitro: {
    prerender: {
      crawlLinks: false
    }
  },
  hooks: {
    async 'nitro:config' (nitroConfig) {
      if (nitroConfig.dev) { return }

      const routes = await dynamicRoutes()
      nitroConfig.prerender.routes = [...routes]
    }
  },
  target: 'static',
```

This it what I have at the moment. But it looks like there were still API calls yesterday but can't reproduce it at the moment.
https://craftcms-baukasten-headless.vercel.app/

**Reply by TheAlexLichter** [reply]
@davidhellmann PR welcome!

You can remove `target: 'static'` here. Other than that, LGTM.
Be aware that "full static" is experimental, see https://github.com/nuxt/framework/issues/6411 for more

**Reply by davidhellmann** [reply]
I think my knowledge here is to bad to write documentation :D

**Reply by TheAlexLichter** [reply]
@davidhellmann Then feel free to raise an issue or new discussion with the questions that are unanswered by the current documentation. This will help as well ☺️

**Reply by DenFin** [reply]
@manniL I'm still a bit confused. I thought there are three modes: spa, ssr and ssg, which could be configured like this:

`ssr: false` -> spa mode (data is fetched on client side)
`ssr: true `-> ssr, data is fetched on server during run time
`ssr: true` & `target: 'static'` -> ssg, data is fetched during build time, static site is generated

Do I understand it correctly, that when I remove `target: 'static'` from `nuxt.config` and deploy my app and use `nuxi generate` it will generate a static site anyways?

**Reply by TheAlexLichter** [reply]
@DenFin That is more or less how it was handled in Nuxt 2.

Now you have:

## SSR

`ssr: false`	-> No actual content fetched on the server, will always deliver "almost empty" HTML file with loading spinner or similar
`ssr: true` 	-> Fetch content on the server and Include content in the HTML

## Build commands

`nuxi generate`	-> Generate a static site during build time (with fallback files for SPA-like behavior for e.g. 404s) - usually "Classic SSG" or "prerendering" 
`nuxi build` 		-> Build the site via "dynamic" SSR (and e.g. for a specific platform like Vercel, Cloudflare workers and so on). This will also cover hybrid approaches in the future.

PS: You can already mix "dynamic" SSR with prerendering by specifying`nitro.prerender.routes`. These will be generated statically while other routes will be generated on the fly.

PPS: Be aware, that all of that only applies to the initial request. From then on, it is a good old SPA.

**Reply by SebbeJohansson** [reply]
Some more information on prerendering, settings so that nitro prerenders routes and activating crawler makes it so that you don't need to send in routes unless you have routes that are not on the page initial load. I send in the root, and it crawls all links on the startpage and in the menu.

setup in nuxt.config.ts:
```ts
  nitro: {
    prerender: {
      crawlLinks: true,
      routes: ['/', '/404.html', '/200.html'],
    },
  },
```
as of RC12 I don't need to do anything special in my project besides that setting since it will automatically save the payloads for my fetching during ssr generation.

My completely working project that is in production deployed to netlify that uses this method:
https://github.com/SebbeJohansson/sebbejohansson-front

**Reply by ricky11** [reply]
> It needs much more clearer documentation here I think.
> 
> ```js
>   nitro: {
>     prerender: {
>       crawlLinks: false
>     }
>   },
>   hooks: {
>     async 'nitro:config' (nitroConfig) {
>       if (nitroConfig.dev) { return }
> 
>       const routes = await dynamicRoutes()
>       nitroConfig.prerender.routes = [...routes]
>     }
>   },
>   target: 'static',
> ```
> 
> This it what I have at the moment. But it looks like there were still API calls yesterday but can't reproduce it at the moment. https://craftcms-baukasten-headless.vercel.app/

what is in your dynamicRoute function? would like to know how to get ssg for dynamic routes to work.

**Reply by davidhellmann** [reply]
@ricky11 
```ts
import axios from 'axios'
import { configSites } from '../config/sites';

export const dynamicRoutes = async () => {
  const finalRoutes: string[] = []
  const {data: {data: {entries}}} = await axios({
    url: `${process.env.API_BASE_URL}/api` || '',
    headers: {
      Authorization: `Bearer ${process.env.GRAPHQL_TOKEN}`,
    },
    method: 'post',
    data: {
      query: `
          {
            entries(section: ["home", "pages", "news"], site: ["default", "defaultDe", "defaultFr"]) {
              uri
              siteId
            }
          }
          `,
    },
  })

  entries.forEach((entry) => {
    let { siteId, uri } = entry

    const matchingSite = configSites.find(x => x.siteId === siteId);
    const locale = matchingSite.urlParameter.length > 0 ? `/${matchingSite.urlParameter}/` : '/';

    finalRoutes.push(`${locale}${uri.replace('__home__', '')}`)
  })

  return finalRoutes
}
```

If you have questions, let me know. 


edit:

sites.ts
```ts
export interface IConfigSites {
    handle: string,
    siteId: number,
    language: string,
    urlParameter?: string,
    urlParameterTrailingSlash?: string,
    primary: boolean
}

export const configSites:IConfigSites[] = [
  {
    handle: 'default',
    siteId: 1,
    language: 'en',
    urlParameter: '',
    urlParameterTrailingSlash: '',
    primary: true,
  },
  {
    handle: 'defaultDe',
    siteId: 4,
    language: 'de',
    urlParameter: 'de',
    urlParameterTrailingSlash: 'de/',
    primary: false,
  },
  {
    handle: 'defaultFr',
    siteId: 5,
    language: 'fr',
    urlParameter: 'fr',
    urlParameterTrailingSlash: 'fr/',
    primary: false,
  }
]
```

**Reply by ricky11** [reply]
Thanks, for this...I thought that nuxt pre-rendering would just handle all this. The headless chrome that is pre-rendering when you `generate` should be able to search for links within the application at build time, and fetch all the data, shove them in a payload.js and build all the paths... is this not something that nuxt does on its own without writing custom hooks and functions?

The docs say it will scan the site and ssg all the links, or i can specific them manually as such :
` prerender: {
         routes: ['/']
    }` 
    
The problem we are having , only my top level links are being generated, no nested [dynamic] child routes are being pre-rendered. Would this be something that nuxt2 would  and nuxt3 is still working on?

**Reply by davidhellmann** [reply]
Sure and this also works but just for sites that are linked somewhere. For me, it's more robust to handle it this way to be sure to have all sites generated. 

And I have not really experience with nuxt2 here.

**Reply by ricky11** [reply]
I can manually prerender my routes if I set the absolute path as such :

```
routes: ['/','/products',
        '/products/blitzen-embroidery-christmas-tablecloth-and-napkins',
        '/products/applique-poinsettia-table-runner']
    }
```
   
    This will work and pre-renders, BUT I have about 100 of this routes, and I need to set it dynamically using /[brackets]
    
    but this does not work, there is nothing in the docs that discusses this.
    
    So my question if anybody knows is 
    1) if I `npm run generate` and the crawler parses my site why cant it find all the nuxt-links and parses them?
    2) If that doesn't work, then how do I set pre-render dynamic links in nuxt.config file?
    
    Thank you.

**Reply by davidhellmann** [reply]
1. I'll try that with subpages out later 
2. like my approach? or whats wrong with that?

**Reply by ricky11** [reply]
I see what your function is doing, which is  to fetch all the possible dynamic routes, put them in a string and then in to the arrays's finalRoutes.   Thats good and all, but my question to the nuxt team is why is this needed for all routes,  shouldn't the nitro crawler itself be able to see the nuxt-links and generate static pages on its own?

**Reply by danielroe** [reply]
It can indeed, as long as you have links in the rendered HTML.

If you are running `nuxi build` you can enable the crawler with `nitro.prerender.crawlLinks`. It's enabled by default with `nuxi generate`.

**Reply by ricky11** [reply]
Thanks @danielroe.  We are using `nuxi generate` and it seems, I fail to  generate static content for dynamic pages which are in [[double brackets]] / optional. 

i.e. 
`/pages
  - /products\[category]\[[theme]]\[slug]` <-- cant generate static files (no error shown)
  `-/products\[category]\index.vue`  <--this works with payload
  
  in the above pages structure, only the category gets generated, but the other pages do not.
  When viewing the live site, all the routes work as expected, and network tab shows that payload is being fetched for all the [slug] pages.  However when page is refreshed  (for non rendered links) an  error in console shows  : `TypeError: Cannot read properties of null (reading 'data')`  -- this only happens on page refresh, not when navigating.
  
  the category index page does have `nuxt-links`  or `generateTo` links to the [[theme]]\[slug] and they work when we npm run serve the pages in dev mode.. Im unsure as to why the crawler cant find these links in the pages when it pre-renders them.
  
  Only if i specify the routes like so in nuxt.config then the data is pre-rendered OK.
  
  routes: ['/','/products',
        '/products/christmas/blitzen-embroidery-christmas-tablecloth-and-napkins',
        '/products/fall/applique-poinsettia-table-runner']
    }
    
    But I have 300 of those links which are dynamic.  i am sure i am doing something wrong..

**Reply by danielroe** [reply]
calling `navigateTo` on button click won't do anything - you need an actual `<a href>` rendered in the html. Also, you probably shouldn't have `ssr: false` enabled as this will not output any crawlable  HTML.

**Reply by ricky11** [reply]
🤯 mind blown.. ofcourse, why did i think that navigateTo would produce anchor links! I changed them to nuxt-links and boom , there pre-rendered.  some questions to follow..!  (for newbs, this would be nice to note in the docs)

**Reply by davidhellmann** [reply]
Hm... interesting...

I've tried it also out with the following config:

```ts
  nitro: {
    prerender: {
      crawlLinks: true,
      routes: ['/']
    }
  },
```

it creates just three pages (homepage in three different languages). 
But there is an navigation with links:
![CleanShot 2022-10-04 at 18 24 04](https://user-images.githubusercontent.com/764292/193873342-8c38dc02-751c-4bfd-82bf-9cd7a913b764.png)

I'm fine with the hook approach then I think it's a better solution because not all pages are linked within the content. But would be great if that would works too.

And congrats @ricky11 :)

**Reply by SebbeJohansson** [reply]
@davidhellmann That is weird. That setup works for me in my project.

**Reply by davidhellmann** [reply]
Can you share your config?

**Reply by SebbeJohansson** [reply]
@davidhellmann Maybe you are using i18n or something that makes it not work? I haven't gotten around to trying i18n with this yet.

Here is the config for my project: https://github.com/SebbeJohansson/sebbejohansson-front/blob/main/nuxt.config.ts

**Comment by DenFin** [comment]
I am trying to fully understand how Static Side Generation works in Nuxt 3. By fully understanding I mean, not just knowing how to configure my app, but knowing what happens between executing `nuxi generate` and having the static HTML files.

So here is an assumption and two questions:

**Assumption 1**
Static Side Generation means, that when a user initially loads/visits a page he gets served static HTML. This is how a search engine crawler would see the site. On client-side navigation we are back in SPA mode, which means that the HTML gets hydrated via the `_payload.js` file, which contains all the page data. That means if the data fetching composables of Nuxt 3 are used 'correctly' no HTTP calls are sent to fetch data, but instead the data comes from the `_payload.js`

**Question 1**
How do nitro and rollup play together in this process? Is nitro used to run rollup? And rollup handles the generation of static HTML files?

**Question 2**
I would really love to take a deep dive into this topic and fully understand how this works by looking at the source code.
I started by looking at the generate command in `framework/packages/nuxi/src/command/generate.ts` and tried to navigate through the code. I ended up in the `build.ts` file and did not really understand what was happening there. Where does the generation of the static HTML files happen? Which parts of the code could 'demystify' the process?

**Comment by dimasxp** [comment]
how to generate all  payloads.js files with unique hash in full static mode.
I have a problem with cloudflare cache, all the payloads are cached.
any suggestion ?

**Reply by ricky11** [reply]
I had this issue but found out that it was my browser that was caching the data. Once I loaded the site in a new profile or ingognito I saw fresh data.   so I would check that first.   also try to purge all cache.   

my guess is that you want to hash your file name.   I would like to know that too.

**Reply by dimasxp** [reply]
yes, purge cache is solve the problem for me, but how to deal with it on production?

**Reply by pislarm** [reply]
Did you maybe find a working solution? I'm experiencing the same issue on a heavy traffic website. Can't seem to find how to hash payload files.

**Comment by andystevensname** [comment]
It seems like Nuxt only generates half of my content files.

My content looks like this:
```
/content
  /blog
    - a-blog-post.md
  - about.md
  - another-top-level-page.md
```

My pages are set up to render that content with `[slug].vue` files:

```
/pages
  - index.vue (/)
  - [slug.vue] (/about, /another-top-level-page)
  /blog
    - index.vue (/blog)
    - [slug].vue (/blog/a-blog-post
    /tags
      - index.vue (/blog/tags)
      - [slug].vue (/blog/tags/a-tag)
```

If I don't declare my routes in my `nuxt.config`, Nuxt will generate all the content in `/blog` but none of the top-level pieces of content like `/about` and `/another-top-level-page`, so I have to declare those routes. It's not the end of the world because I don't have that much other top-level content, but I'm unsure why that's happening. Anyone have any ideas? [Here is the repo in question.](https://github.com/andystevensname/andystevens.name).

**Reply by danielroe** [reply]
Nuxt will crawl any links you have on the page but otherwise doesn't know they exist. So if you have dynamic pages you will need to either include a page linking to them or add them manually. (This may be something that content can help do - or already does - in the document-driven mode. cc: @Tahul)

**Comment by chinenvinicius** [comment]
Hi im using nuxt 3 to generate my static site. so i decided to use directus and call api then use nuxt generate command. however, every time i my directus host is off , the content disappear. what should i do? isnt it supporsed to be pre rendered?

**Reply by ricky11** [reply]
your cms, whatever it may be should be always up and running whenever you `npm run generate`
i am not sure what you mean by your directus is off, is it self hosted or cloud hosted.

**Reply by chinenvinicius** [reply]
@ricky11 , hi im doing sel hositing, hosting locally on my pc.  so i keep the cms api running when im doing npm run generate. im just trying to see if the content wont disappear when cms is not running.  i thought npm run generate only need to do one api call and the content will generated on html, so i dont need to do api call all the time. after the npm run generate finished, i stopped my cms to see if the content stay, but it doesnt.

**Reply by ricky11** [reply]
when you generate, do you see the _payload.js and index.html file being generated in the ./output folder?   

oh you mean when your api is off, your website still tries to connect to the api to get data?  this means that it is not cached in the _payload.js file, you can check if the data is there... are you using useFetch or useAsyncFetch did you give it a unique key name?

**Reply by chinenvinicius** [reply]
@ricky11 ooh i didnt know i have to put a key on v-for loop. so i put :key="" on v-for loop, i notice the data was on payload.js. then i tried to check by using npm run preview, the data was there even when api was not running. however when did git push and deployed on netlify , i see that browser is making request to the api.

`const url = "http://localhost:8055/items/Articles";
const  {data} = await useAsyncData("key", async ()=>{
  return $fetch(url)
});`

this is the code im using

**Reply by ricky11** [reply]
I am using the useFetch wrapper with soemthing like this : const {data,pending,error,refresh} = await useFetch(url,{key:'tradeshow'})
i think key should be an object with {key : 'xxxx'}
useFetch has some nice defaults, you may need to look at the docs if you are using useAsycnData.

chcek the netlify logs are the payloads being generated.

**Reply by chinenvinicius** [reply]
@ricky11 i see. i checked but the payload is not being generated . 
[](https://chic-starburst-d53b3a.netlify.app/)
here is the link of simple project im doing. the payload is not showing on dev tools.

**Comment by kuzroman** [comment]
can somebody explain? i generate nuxt 3 app and it work on my local http node server, but if i push it to github pages, some files have 404 error & js failed. https://kuzroman.github.io/nuxt3/

**Reply by SebbeJohansson** [reply]
Which app do you want to go to? From what i can tell there is a redirect to https://nuxt3-ibvd.onrender.com/. Going to /nuxt3/ doesnt work, but im not sure if that route should work?

**Comment by surgiie** [comment]
In nuxt 2, you were able to specify `payload` in a route object during a `nuxt generate` to provide the data for the static html file:

https://v2.nuxt.com/docs/configuration-glossary/configuration-generate/#speeding-up-dynamic-route-generation-with-payload


I dont see any docs on how to accomplish this in nuxt 3, passing this payload during a generate is not rendering this data as it was before.


Am i missing something?
